util/topgen/merge.py

```diff
@@ -852,127 +851,187 @@ def amend_reset_request(topcfg: OrderedDict,
         topcfg["inter_module"]["connect"]))
 
 
-def amend_pinmux_io(top: OrderedDict, name_to_block: Dict[str, IpBlock]):
-    """ Check dio_modules/ mio_modules. If not exists, add all modules to mio
-    """
-    pinmux = top["pinmux"]
+def append_io_signal(temp: Dict, sig_inst: Dict):
+    '''Appends the signal to the correct list'''
+    if sig_inst['type'] == 'inout':
+        temp['inouts'].append(sig_inst)
+    if sig_inst['type'] == 'input':
+        temp['inputs'].append(sig_inst)
+    if sig_inst['type'] == 'output':
+        temp['outputs'].append(sig_inst)
+
+
+def get_index_and_incr(ctrs: Dict, pad: str, io_dir: str):
+    '''Get correct index counter and increment'''
+
+    if pad != 'muxed':
+        pad = 'dedicated'
+
+    if io_dir in 'inout':
+        result = ctrs[pad]['inouts']
+        ctrs[pad]['inouts'] += 1
+    elif pad == 'muxed':
+        # For MIOs, the input/output arrays differ in RTL
+        # I.e., the input array contains {inputs, inouts}, whereas
+        # the output array contains {outputs, inouts}.
+        if io_dir == 'input':
+            result = ctrs[pad]['inputs'] + ctrs[pad]['inouts']
+            ctrs[pad]['inputs'] += 1
+        elif io_dir == 'output':
+            result = ctrs[pad]['outputs'] + ctrs[pad]['inouts']
+            ctrs[pad]['outputs'] += 1
+        else:
+            assert(0)  # should not happen
+    else:
+        # For DIOs, the input/output arrays are identical in terms of index layout.
+        # Unused inputs are left unconnected and unused outputs are tied off.
+        if io_dir == 'input':
+            result = ctrs[pad]['inputs'] + ctrs[pad]['inouts']
+            ctrs[pad]['inputs'] += 1
+        elif io_dir == 'output':
+            result = ctrs[pad]['outputs'] + ctrs[pad]['inouts'] + ctrs[pad]['inputs']
+            ctrs[pad]['outputs'] += 1
+        else:
+            assert(0)  # should not happen
+
+    return result
 
-    if "dio_modules" not in pinmux:
-        pinmux['dio_modules'] = []
 
-    # list out dedicated IO
-    pinmux['dio'] = []
-    for e in pinmux["dio_modules"]:
-        # Check name if it is module or signal
-        mname, sname = lib.get_ms_name(e["name"])
+def amend_pinmux_io(top: Dict, name_to_block: Dict[str, IpBlock]):
+    """ Process pinmux/pinout configuration and assign available IOs
+    """
+    pinmux = top['pinmux']
+    pinout = top['pinout']
+    targets = top['targets']
+
+    temp = {}
+    temp['inouts'] = []
+    temp['inputs'] = []
+    temp['outputs'] = []
+
+    for sig in pinmux['signals']:
+        # If the signal is explicitly named.
+        if '.' in sig['name']:
```

Comment:
As commented in the top_earlgrey.hjson file -- let's simplify parsing by using each field for exactly one purpose, which will get rid of the "it could be a signal, or could be a module" if/else block here.

Reply:
I have changed the datastructure along the lines of what you proposed. 
This does simplify the parsing and name checking a bit.
Note however that we still have to distinguish the case where the port is explicitly named from the case where it is not (the latter acts as a wildcard).